src/librustc_mir/interpret/intrinsics.rs

```diff
@@ -441,9 +441,11 @@ impl<'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
         // We cannot overflow i64 as a type's size must be <= isize::MAX.
         let pointee_size = i64::try_from(self.layout_of(pointee_ty)?.size.bytes()).unwrap();
         // The computed offset, in bytes, cannot overflow an isize.
-        let offset_bytes = offset_count
-            .checked_mul(pointee_size)
-            .ok_or(err_ub_format!("inbounds pointer arithmetic: overflow computing offset"))?;
+        let offset_bytes =
+            offset_count.checked_mul(pointee_size).ok_or(err_ub!(PointerArithOverflow))?;
+        if offset_bytes < self.machine_isize_min() || offset_bytes > self.machine_isize_max() {
+            throw_ub!(PointerArithOverflow)
+        }
```

Comment:
Is there any observable change in behavior from this extra check? If not, there should be at least a comment explaining why we do it.

I am not entirely sure it is worth it, given that `ptr_signed_offset` already has to check this, too.

Reply:
So `ptr_signed_offset` does not catch the `isize` overflow if `offset_bytes` is between `isize::MAX` and `usize::MAX`. The UB is still caught in that case, but just in the memory check below

I'll add a comment explaining why this check exists.

Reply:
> So ptr_signed_offset does not catch the isize overflow if offset_bytes is between isize::MAX and usize::MAX.

That seems like a serious footgun, good catch!

If we decide that it is up to the caller to ensure that the input to these ptr offset methods is in the range of the target pointer type, could you add appropriate assertions? Alternatively we should make those methods properly detect overflow even for these inputs.

Reply:
Like a `debug_assert` in libcore?

Reply:
In libcore? Now I am very confused.

I was more thinking of librustc_middle/mir/interpret/pointer.rs. And I'd make it proper `assert!` unless we have data that it actually slows down anything.

Reply:
I just added checks to `PointerArithmetic`'s `offset` and `signed_offset` methods to just detect overflow of the offset in these cases (and return `PointerArithOverflow`). This should make doing the right thing easier and means we don't need this check in `librustc_mir`.